File: step4/a_star.py
#!/usr/bin/python3
def reconstruct_path(came_from, current):
    total_path = [(current, 'goal')]

    while current in came_from:
        current, action = came_from[current]
        total_path.insert(0, (current, action))

    return total_path


INF = float('inf')


def tuple_diff(a, b):
    return [(field, getattr(a, field), getattr(b, field))
            for field in b._fields
            if getattr(a, field) != getattr(b, field) and
            getattr(a, field) != ... and getattr(b, field) != ...]


def a_star(start, goal, neighbors, h):
    open_set = [start]
    came_from = {}

    # Cost of path from start node to n
    g_score = {}
    g_score[start] = 0

    # Estimated cost of path from start node through n to goal
    # This is an estimate of the total path cost
    f_score = {}
    f_score[start] = h(start, goal)

    while len(open_set) > 0:
        open_set.sort(key=lambda id: f_score.get(id, INF))
        current = open_set.pop(0)

        if len(tuple_diff(current, goal)) == 0:
            return reconstruct_path(came_from, current)

        # print()
        # print(current)
        for neighbor, d, action in neighbors(current):
            # print(desc)
            tentative_g_score = g_score.get(current, INF) + d
            if tentative_g_score < g_score.get(neighbor, INF):
                # This path to neighbor is the best one seen so far
                came_from[neighbor] = current, action
                g_score[neighbor] = tentative_g_score
                f_score[neighbor] = g_score[neighbor] + h(neighbor, goal)

                if neighbor not in open_set:
                    open_set.append(neighbor)

File: rustplan/src/a_star.rs
use std::collections::HashMap;

#[derive(Debug)]
pub struct Neighbor<S> {
    state: S,
    cost: f32,
    action: String,
}

impl<S> Neighbor<S> {
    pub fn new(state: S, cost: f32, action: String) -> Neighbor<S> {
        Neighbor {
            state,
            cost,
            action,
        }
    }
}

pub fn a_star<S>(
    start: &S,
    heuristic: &dyn Fn(&S) -> f32,
    neighbors: &dyn Fn(&S) -> Vec<Neighbor<S>>,
) -> Result<Vec<(S, String)>, ()>
where
    S: Clone + PartialEq + Eq + std::hash::Hash + std::fmt::Debug,
{
    let mut open_set: Vec<S> = vec![start.clone()];

    let mut came_from = HashMap::new();

    // Cost of path from start to node
    let mut g_score = HashMap::new();
    g_score.insert(start.clone(), 0.0);

    // Estimated cost of path from start through node to goal
    // This is an estimate of the total path cost.
    let mut f_score = HashMap::new();
    f_score.insert(start.clone(), heuristic(&start));

    while open_set.len() > 0 {
        open_set.sort_by(|a, b| {
            f_score
                .get(a)
                .unwrap_or(&f32::INFINITY)
                .partial_cmp(f_score.get(b).unwrap_or(&f32::INFINITY))
                .unwrap()
        });

        let current = &open_set.remove(0);

        if heuristic(current) == 0.0 {
            return Ok(reconstruct_path(came_from, current));
        }

        for neighbor in neighbors(current) {
            let tentative_g_score = g_score.get(current).unwrap_or(&f32::INFINITY) + neighbor.cost;
            if tentative_g_score < *g_score.get(&neighbor.state).unwrap_or(&f32::INFINITY) {
                // This path to the neighbor is the best one seen so far
                came_from.insert(
                    neighbor.state.clone(),
                    (current.clone(), neighbor.action.to_owned()),
                );
                g_score.insert(neighbor.state.clone(), tentative_g_score);
                f_score.insert(
                    neighbor.state.clone(),
                    tentative_g_score + heuristic(&neighbor.state),
                );

                if !open_set.contains(&&neighbor.state) {
                    open_set.push(neighbor.state.clone())
                }
            }
        }
    }

    return Err(());
}

fn reconstruct_path<S>(came_from: HashMap<S, (S, String)>, current: &S) -> Vec<(S, String)>
where
    S: Clone + Eq + std::hash::Hash,
{
    let mut total_path: Vec<(S, String)> = vec![(current.clone(), "goal".to_owned())];
    let mut current = current;

    while let Some((state, action)) = came_from.get(current) {
        total_path.insert(0, (state.clone(), action.to_owned()));
        current = state;
    }

    total_path
}
